cmdutils/switches/switches.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2021 OnMetal authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package switches provides new type that implements flag.Value interface -- Switches.
+// It can be used for enabling/disabling controllers/webhooks in your controller manager.
+package switches
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	defaultValue  = "*"
+	disablePrefix = "-"
+)
+
+type Switches struct {
+	settings map[string]bool
+}
+
+// New creates an instance of Switches
+func New(settings []string) *Switches {
+	s := &Switches{
+		settings: make(map[string]bool),
+	}
+	s.setSettings(settings)
+	return s
+}
+
+// Disable prepends disablePrefix prefix to a controller name
+func Disable(name string) string {
+	return disablePrefix + name
+}
+
+func (s *Switches) String() string {
+	return fmt.Sprintf("%v", s.settings)
+}
+
+func (s *Switches) Set(val string) error {
+	s.setSettings(strings.Split(val, ","))
+	return nil
+}
+
+// Enabled checks if controller is enabled
+func (s *Switches) Enabled(name string) bool {
+	return s.settings[name]
+}
+
+// All returns names of all controllers set in settings
+func (s *Switches) All() (names []string) {
+	for k := range s.settings {
+		names = append(names, k)
+	}
+
+	return
+}
+
+// DisabledByDefault returns names of all disabled controllers
+func (s *Switches) DisabledByDefault() (names []string) {
+	for k, enabled := range s.settings {
+		if !enabled {
+			names = append(names, k)
+		}
+	}
+
+	return
+}
+
+func (s *Switches) Type() string {
+	return "Switches"
+}
+
+func (s *Switches) setSettings(settings []string) {
+	if len(settings) == 1 && settings[0] == "" {
+		return
+	}
+
+	var isDefault bool
+	for _, v := range settings {
+		if v == defaultValue {
+			isDefault = true
+			break
+		}
+	}
+
+	if !isDefault {
+		s.settings = make(map[string]bool)
```

Comment:
Why are all settings cleared in case no `*` is contained in the settings? `*` should just (re)enable all values that are enabled by default.

Reply:
So, for example, if by default we have `controller-a,-controller-b`. And user passes:
1) `controller-b` -- it enables controller-b
2) `*,controller-b` -- it also enables controller-b

Am i correctly understand you? Then what's the point of `*`? Flags can be set only once, at the startup of service.

Reply:
The point of `*` is that you immediately reference the set of controllers that are enabled by default and can add / remove on top of that.
Consider having 100 controllers, 50 of them enabled by default and 50 not. With the controller switches, it's now possible to
* Only enable e.g. 3 specific controllers by passing `a,b,c`, leaving 97 controllers not enabled.
* Enable all controllers enabled by default + 3 chosen others by passing `*,x,y,z`
* Enable all controllers enabled by default but disabling some of them by passing `*,-a,-b`

Reply:
Ah, i see. That's exactly what current implementation does. If "*" not specified -- we clear all default settings and override them with new ones.

I guess, you're confusion comes from the fact that we actually store default serttings and user provided settings in the same data structure. So, default settings are not stored anywhere additionally. Because that's not needed, since `switches` can be used only once at the service start up.

Reply:
There some important things here:
* If the user does not specify anything, `*` should be used.
* As soon as the users specifies something, it should be overrriden.
* Validation for the controllers named should still be present.